Simplenote/NoticePresenter.swift

```diff
@@ -0,0 +1,170 @@
+    }
+
+    @objc
+    private func dismissAnimation(noticeView: NoticeView, completion: @escaping () -> Void) {
+        UIView.animate(withDuration: UIKitConstants.animationLongDuration) {
+            noticeView.alpha = .zero
+        } completion: { (_) in
+            self.noticeView?.removeFromSuperview()
+            self.containerView.removeFromSuperview()
+            self.noticeView = nil
+            self.timer = nil
+            completion()
+        }
+    }
+
+    // MARK: Keyboard Obserbers
+    //
+    func keyboardDidChangeFrame(beginFrame: CGRect?, endFrame: CGRect?, animationDuration: TimeInterval?, animationCurve: UInt?) {
+        guard let endFrame = endFrame,
+              let animationCurve = animationCurve,
+              let animationDuration = animationDuration else {
+            return
+        }
+        updateKeyboardHeight(with: endFrame)
+        animateNoticeToNewKeyboardLocation(frame: endFrame, curve: animationCurve, animationDuration: animationDuration)
+    }
+
+    func keyboardWillChangeFrame(beginFrame: CGRect?, endFrame: CGRect?, animationDuration: TimeInterval?, animationCurve: UInt?) {
+        guard let endFrame = endFrame,
+              let animationCurve = animationCurve,
+              let animationDuration = animationDuration else {
+            return
+        }
+        updateKeyboardHeight(with: endFrame)
+        animateNoticeToNewKeyboardLocation(frame: endFrame, curve: animationCurve, animationDuration: animationDuration)
+    }
+
+    private func updateKeyboardHeight(with frame: CGRect) {
+        let windowFrame = getWindowFrame()
+        keyboardHeight = frame.intersection(getWindowFrame()).height
+        keyboardFloats = frame.maxY < windowFrame.height
+    }
+    private func animateNoticeToNewKeyboardLocation(frame: CGRect, curve: UInt, animationDuration: TimeInterval) {
+        if !isPresenting {
+            return
+        }
+
+        let animationOptions = UIView.AnimationOptions(arrayLiteral: .beginFromCurrentState, .init(rawValue: curve))
+
+        noticeVariableConstraint?.constant = makeBottomConstraintConstant()
+        UIView.animate(withDuration: animationDuration, delay: .zero, options: animationOptions) {
+            self.containerView.layoutIfNeeded()
+        }
+    }
+
+    func startListeningToKeyboardNotifications() {
+        keyboardNotificationTokens = addKeyboardObservers()
+    }
+
+    func stopListeningToKeyboardNotifications() {
+        guard let tokens = keyboardNotificationTokens else {
+            return
+        }
+
+        removeKeyboardObservers(with: tokens)
+        keyboardNotificationTokens = nil
+    }
+}
+
+extension NoticePresenter {
+    // Convenience method to fetch current key window
+    //
+    private func getKeyWindow() -> UIWindow? {
+        return UIApplication.shared.windows.first(where: { $0.isKeyWindow })
+    }
+
+    private func getWindowFrame() -> CGRect {
+        return getKeyWindow()?.frame ?? .zero
+    }
```

Comment:
If possible method names should not have `get*` in them. 
`keyWindow` and `windowFrame` would match current naming conventions better.